### PR TITLE
Add module to set MOTD banner text for Pure Storage FlashArray

### PR DESCRIPTION
**Pull request by sdodsley**
##### SUMMARY
New module to set or delete the MOTD banner text that is shown when logging in via SSH or the GUI

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
purefa_banner.py

### CODE REVIEW

**Comment by genegr** [comment]
shipit

File: lib/ansible/modules/storage/purestorage/purefa_banner.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                           required_if=required_if,
+                           supports_check_mode=False)
+
+    state = module.params['state']
+    array = get_system(module)
+    current_banner = array.get(banner=True)['banner']
+    if current_banner != module.params['banner'] and state == 'present':
+        set_banner(module, array)
+    elif state == 'absent':
+        delete_banner(module, array)
```
Comment (resmo) [inline lib/ansible/modules/storage/purestorage/purefa_banner.py:110]: this seems not to be idempotent, my suggestion
~~~python

    if current_banner != module.params['banner']
        if state == 'present':
            set_banner(module, array)
        else:
            delete_banner(module, array)
~~~

File: lib/ansible/modules/storage/purestorage/purefa_banner.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        module.fail_json(msg='Failed to set MOTD banner text')
+
+    module.exit_json(changed=changed)
+
+
+def delete_banner(module, array):
+    """Delete MOTD banner text"""
+    changed = False
+    try:
+        array.set(banner="")
```
Comment (resmo) [inline lib/ansible/modules/storage/purestorage/purefa_banner.py:82]: catching a low hanging fruit

~~~python
        if not module.check_mode:
            array.set(banner="")
        changed = True
~~~

File: lib/ansible/modules/storage/purestorage/purefa_banner.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pure import get_system, purefa_argument_spec
+
+
+def set_banner(module, array):
+    """Set MOTD banner text"""
+    changed = False
+
+    try:
+        array.set(banner=module.params['banner'])
```
Comment (resmo) [inline lib/ansible/modules/storage/purestorage/purefa_banner.py]: added check mode and check for empty banner as this is unexpected --> use case state=absent
~~~python
        if not module.params['banner']:
            module.fail_json(msg='Invalid MOTD banner given')

        if not module.check_mode:
            array.set(banner=module.params['banner'])
        changed = True
~~~

File: lib/ansible/modules/storage/purestorage/purefa_banner.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    argument_spec.update(dict(
+        banner=dict(type='str', default="Welcome to the machine..."),
+        state=dict(type='str', default='present', choices=['present', 'absent']),
+    ))
+
+    required_if = [('state', 'present', ['banner'])]
+
+    module = AnsibleModule(argument_spec,
+                           required_if=required_if,
+                           supports_check_mode=False)
```
Comment (resmo) [inline lib/ansible/modules/storage/purestorage/purefa_banner.py:100]: set supports_check_mode=True after the check mode implementation

**Review by resmo — CHANGES_REQUESTED** [review]
see comments inline, mostly low hanging fruits: check mode support and missing idempotency in state=absent

File: lib/ansible/modules/storage/purestorage/purefa_banner.py (hunk 4 above)
Comment (sdodsley) [reply lib/ansible/modules/storage/purestorage/purefa_banner.py:100]: @resmo All revisions done as requested. Could you re-review please?

File: lib/ansible/modules/storage/purestorage/purefa_banner.py (hunk 1 above)
Comment (resmo) [reply lib/ansible/modules/storage/purestorage/purefa_banner.py:110]: this doesn't look good to me yet, but I see my previous suggestion looks wrong too. How about the following?
~~~python 
    # set banner if empty value or value differs 
    if state == 'present' and (not current_banner or current_banner != module.params['banner']):
        set_banner(module, array)
   # clear banner if it has a value
    elif state == 'absent' and current_banner:
        delete_banner(module, array)
~~~

Comment (sdodsley) [reply lib/ansible/modules/storage/purestorage/purefa_banner.py:110]: OK - I think we are good with this version

**Review by resmo — APPROVED** [review]
shipit

**Pull request merged**